### DISCUSSION TITLE
ORG Recognizer - Rule-based - Difficulties to set up

### DISCUSSION THREAD

**Discussion by LSD-98** [microsoft/presidio#1121] [Q&A] [answered]
Hello all,

Sorry if I am not posting this at the right place, I am quite new to Git & Python.
I am trying to anonymize pdf documents containing names, addresses and **company names**. 
Thus, I am trying to create and add a new recognizer to detect company names. I am taking [this example](https://microsoft.github.io/presidio/tutorial/03_rule_based/) from the documentation as a baseline.

```
from typing import List
from presidio_analyzer import EntityRecognizer, RecognizerResult
from presidio_analyzer.nlp_engine import NlpArtifacts


class OrgRecognizer(EntityRecognizer):
    expected_confidence_level = 0.7  # expected confidence level for this recognizer
    def load(self) -> None:
        """No loading is required."""
        pass

    def analyze(
        self, text: str, entities: List[str], nlp_artifacts: NlpArtifacts
    ) -> List[RecognizerResult]:
        """
        Analyzes test to find tokens which represent organizations.
        """
        results = []
        # iterate over the spaCy tokens, and call `token.like_num`
        for token in nlp_artifacts.tokens:
            if token.like_num:
                result = RecognizerResult(
                    entity_type='ORG',
                    start=token.idx,
                    end=token.idx + len(token),
                    score=self.expected_confidence_level,
                )
                results.append(result)
        return results

# Instantiate the new OrgRecognizer:
new_org_recognizer = OrgRecognizer(supported_entities=["ORG"])

from presidio_analyzer import AnalyzerEngine

text3 = "Léo works at Microsoft and Apple at the same time !"
analyzer = AnalyzerEngine()
analyzer.registry.add_recognizer(new_org_recognizer)
analyzer.get_supported_entities()

results1 = analyzer.analyze(text=text3, language="en")
print("Results:")
print("\n".join([str(res) for res in results1]))

```
It results in the following :  `type: PERSON, start: 0, end: 3, score: 0.85` and does not detect company names.
I thought it would work as "ORG" is an entity_type of spacy, but I think I have a problem within `def analyze`...

Can someone walk me through how to set up such a recognizer? Also, what should i do if i wanted to use a different NLP model for this recognizer (for example to support french instead of English ?)

Thanks in advance for your time and thanks for the work you put in on this package, it's really helpful.

**Comment by LSD-98** [comment, marked as answer]
Found the solution by exploring the code files, and notably `spacyrecognizer.py`

Here is the code that works in case it helps anyone :
```
from presidio_analyzer import AnalyzerEngine
from presidio_analyzer.predefined_recognizers import SpacyRecognizer

# Setting up Org_Recognizer & adding it to the analyzer
Entities = ["ORGANIZATION"]
OrgRecognizer = SpacyRecognizer(supported_language="en",supported_entities=Entities, ner_strength=0.85)
analyzer=AnalyzerEngine()

analyzer.registry.add_recognizer(OrgRecognizer)
analyzer.get_supported_entities()

# Test
Text = "Mark works at Microsoft and Google and Apple at the same time !"
results = analyzer.analyze(Text, language= 'en')
print(results)
```

Now I would like to process French text, how can I change the NLP engine in the `Org_Recognizer` ?

**Reply by omri374** [reply]
Please see https://microsoft.github.io/presidio/tutorial/05_languages/

Adapting the nlp engine together with your change to the spacy recognizer will allow you to extract org names in French.

**Reply by LSD-98** [reply]
Dear Omri,

Thanks for your reply. I implemented this solution, and it works fine.
